<a href="https://colab.research.google.com/github/AishwaryaJupudi/Rebounce-Applied-AI-and-Analytics-Codes/blob/main/Python_301.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##***Minor Project 3 · Clean, Analyse & Summarise a Real Dataset***

#Part-1 Load and Understand the data

##Import libraries

In [3]:
import pandas as pd
import numpy as np

##Load dataset

In [4]:
url = "https://huggingface.co/datasets/aarav912/online-retail/resolve/main/online_retail.csv"
df = pd.read_csv(url)

##Displaying the first 10 rows

In [5]:
df.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,17850.0,United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,17850.0,United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,17850.0,United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,13047.0,United Kingdom


##Checking the shape of the data

In [6]:
df.shape

(541909, 8)

##Listing the names of columns

In [7]:
df.columns.tolist()

['InvoiceNo',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'UnitPrice',
 'CustomerID',
 'Country']

##Checking the datatype of each column

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


##Counting the missing values in every column

In [9]:
df.isna().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


##Counting the number of duplicate rows

In [10]:
df.duplicated().sum()

np.int64(5268)

The CustomerID and Description columns contain many missing values, while InvoiceDate is stored as text and needs to be converted to datetime. These columns require cleaning before further analysis

#Part - 2 Clean the data

In [11]:
df_clean = df.copy()        # Create a copy of the original dataset for cleaning and analysis

##Missing customer id

Missing CustomerID values were retained because dropping 135080 transactions could affect the analysis and also CustomerID is not required for the sales and country-level analysis.

##Converting invoice date from text to a proper datetime column

In [12]:
df_clean['InvoiceDate'] = pd.to_datetime(df_clean['InvoiceDate'])

In [13]:
df_clean["InvoiceDate"].dtype

dtype('<M8[ns]')

##Removing the duplicate rows

In [14]:
df_clean = df_clean.drop_duplicates()

In [15]:
df_clean.shape       #checking the shape after dropping  duplicates

(536641, 8)

In [16]:
df_clean.duplicated().sum()    #checking the duplicate count after cleaning

np.int64(0)

##Removing invalid transactions


In [17]:
df_clean = df_clean[(df_clean['Quantity']>0) & (df_clean['UnitPrice']>0)]

In [18]:
df_clean.shape

(524878, 8)

##Creating a new Sales column using the formula
    Sales = Quantity × UnitPrice

In [19]:
df_clean['Sales'] = df_clean['Quantity'] * df_clean['UnitPrice']

In [20]:
df_clean.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34


#Part - 3 Numpy Analysis

##Calculating the mean

In [21]:
np.mean(df_clean['Sales'])

np.float64(20.27539886221179)

##Calculating the median

In [22]:
np.median(df_clean['Sales'])

np.float64(9.92)

##Calculating the minimum sales

In [23]:
np.min(df_clean['Sales'])

0.001

##Calculating the maximum sales

In [24]:
np.max(df_clean['Sales'])

168469.6

#Calculating the standard deviation

In [25]:
np.std(df_clean['Sales'])

271.69330748320107

##Rounding the values to a workable number of decimal places

In [26]:
print("Mean:", round(np.mean(df_clean['Sales']),2))
print("Median:",round(np.median(df_clean['Sales']),2))
print("Min:",round(np.min(df_clean['Sales']),3))
print("Max:",round(np.max(df_clean['Sales']),2))
print("Standard Deviation:",round(np.std(df_clean['Sales']),2))

Mean: 20.28
Median: 9.92
Min: 0.001
Max: 168469.6
Standard Deviation: 271.69


The mean sales value (20.28) is considerably higher than the median (9.92), indicating that the sales distribution is right-skewed. This suggests that a small number of very large transactions are pulling the mean upward, while most transactions have relatively lower sales values.

#Part-4 Pandas Filtering

##Filtering the the sales above 500

In [27]:
sales_above_500 = df_clean[df_clean['Sales']>500]
sales_above_500

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
178,536387,79321,CHILLI LIGHTS,192,2010-12-01 09:58:00,3.82,16029.0,United Kingdom,733.44
179,536387,22780,LIGHT GARLAND BUTTERFILES PINK,192,2010-12-01 09:58:00,3.37,16029.0,United Kingdom,647.04
180,536387,22779,WOODEN OWLS LIGHT GARLAND,192,2010-12-01 09:58:00,3.37,16029.0,United Kingdom,647.04
181,536387,22466,FAIRY TALE COTTAGE NIGHTLIGHT,432,2010-12-01 09:58:00,1.45,16029.0,United Kingdom,626.40
182,536387,21731,RED TOADSTOOL LED NIGHT LIGHT,432,2010-12-01 09:58:00,1.25,16029.0,United Kingdom,540.00
...,...,...,...,...,...,...,...,...,...
540071,581459,22197,POPCORN HOLDER,1200,2011-12-08 18:46:00,0.72,17949.0,United Kingdom,864.00
540421,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom,168469.60
540425,581485,20749,ASSORTED COLOUR MINI CASES,84,2011-12-09 09:38:00,6.35,17389.0,United Kingdom,533.40
540908,581492,DOT,DOTCOM POSTAGE,1,2011-12-09 10:03:00,933.17,NaN,United Kingdom,933.17


##Sorting transactions by Sales in descending order and displaying the 10 largest transactions

In [47]:
ten_largest_transactions = df_clean.sort_values('Sales',ascending=False).head(10)
ten_largest_transactions

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
540421,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom,168469.60
61619,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,United Kingdom,77183.60
222680,556444,22502,PICNIC BASKET WICKER 60 PIECES,60,2011-06-10 15:28:00,649.50,15098.0,United Kingdom,38970.00
15017,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,United Kingdom,13541.33
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom,11062.06
173382,551697,POST,POSTAGE,1,2011-05-03 13:46:00,8142.75,16029.0,United Kingdom,8142.75
348325,567423,23243,SET OF TEA COFFEE SUGAR TINS PANTRY,1412,2011-09-20 11:05:00,5.06,17450.0,United Kingdom,7144.72
160546,550461,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-04-18 13:20:00,2.10,15749.0,United Kingdom,6539.40
52711,540815,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-01-11 12:55:00,2.10,15749.0,United Kingdom,6539.40
421601,573003,23084,RABBIT NIGHT LIGHT,2400,2011-10-27 12:11:00,2.08,14646.0,Netherlands,4992.00


#Transactions that come from Germany

In [29]:
df_clean['Country'].unique()                    #Display unique countries to identify the exact country name for filtering

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Finland',
       'Austria', 'Bahrain', 'Israel', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [48]:
Germany_transactions=df_clean[df_clean['Country']== 'Germany']
Germany_transactions

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
1109,536527,22809,SET OF 6 T-LIGHTS SANTA,6,2010-12-01 13:04:00,2.95,12662.0,Germany,17.70
1110,536527,84347,ROTATING SILVER ANGELS T-LIGHT HLDR,6,2010-12-01 13:04:00,2.55,12662.0,Germany,15.30
1111,536527,84945,MULTI COLOUR SILVER T-LIGHT HOLDER,12,2010-12-01 13:04:00,0.85,12662.0,Germany,10.20
1112,536527,22242,5 HOOK HANGER MAGIC TOADSTOOL,12,2010-12-01 13:04:00,1.65,12662.0,Germany,19.80
1113,536527,22244,3 HOOK HANGER MAGIC GARDEN,12,2010-12-01 13:04:00,1.95,12662.0,Germany,23.40
...,...,...,...,...,...,...,...,...,...
541801,581578,22993,SET OF 4 PANTRY JELLY MOULDS,12,2011-12-09 12:16:00,1.25,12713.0,Germany,15.00
541802,581578,22907,PACK OF 20 NAPKINS PANTRY DESIGN,12,2011-12-09 12:16:00,0.85,12713.0,Germany,10.20
541803,581578,22908,PACK OF 20 NAPKINS RED APPLES,12,2011-12-09 12:16:00,0.85,12713.0,Germany,10.20
541804,581578,23215,JINGLE BELL HEART ANTIQUE SILVER,12,2011-12-09 12:16:00,2.08,12713.0,Germany,24.96


##Transactions that have Quantity above 1000

In [49]:
quantity_above_100 = df_clean[df_clean['Quantity']>1000]
quantity_above_100

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Sales
4850,536809,84950,ASSORTED COLOUR T-LIGHT HOLDER,1824,2010-12-02 16:48:00,0.55,15299.0,United Kingdom,1003.20
4945,536830,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,2880,2010-12-02 17:38:00,0.18,16754.0,United Kingdom,518.40
4946,536830,21915,RED HARMONICA IN BOX,1400,2010-12-02 17:38:00,1.06,16754.0,United Kingdom,1484.00
6365,536890,17084R,ASSORTED INCENSE PACK,1440,2010-12-03 11:48:00,0.16,14156.0,EIRE,230.40
16435,537659,22189,CREAM HEART CARD HOLDER,1008,2010-12-07 16:43:00,2.31,18102.0,United Kingdom,2328.48
...,...,...,...,...,...,...,...,...,...
533812,581115,22413,METAL SIGN TAKE IT OR LEAVE IT,1404,2011-12-07 12:20:00,2.75,15195.0,United Kingdom,3861.00
534952,581175,23084,RABBIT NIGHT LIGHT,1440,2011-12-07 15:16:00,1.79,14646.0,Netherlands,2577.60
540070,581458,22197,POPCORN HOLDER,1500,2011-12-08 18:45:00,0.72,17949.0,United Kingdom,1080.00
540071,581459,22197,POPCORN HOLDER,1200,2011-12-08 18:46:00,0.72,17949.0,United Kingdom,864.00


##Do these transactions with quantity above 1000 seem believable?
There are some believable transactions with quantity above 1000, like Paper craft and Popcorn Holder. But some of the high-quantity orders look unusual, so we cannot say for sure whether they are correct or data entry errors. They have to be carefully verified and considered before making any conclusions.

#Part - 5  GroupBy and aggregation

##Building a summary table with each country in a row

In [32]:
country_summary = df_clean.groupby('Country').agg(
    Number_of_transactions= ('Country', 'count'),
    Total_sales=('Sales','sum'),
    Average_sales=('Sales','mean')
).reset_index()

In [33]:
#Rounding the decimal values of Total sales and average sales to 2 decimal places
country_summary['Total_sales']=country_summary['Total_sales'].round(2)
country_summary['Average_sales']=country_summary['Average_sales'].round(2)

In [34]:
country_summary

,Country,Number_of_transactions,Total_sales,Average_sales
0,Australia,1181,138453.81,117.23
1,Austria,398,10198.68,25.62
2,Bahrain,18,754.14,41.90
3,Belgium,2031,41196.34,20.28
4,Brazil,32,1143.60,35.74
5,Canada,151,3666.38,24.28
6,Channel Islands,747,20440.54,27.36
7,Cyprus,603,13502.85,22.39
8,Czech Republic,25,826.74,33.07
9,Denmark,380,18955.34,49.88


##Five countries that generated the highest total sales

In [50]:
top_five_countries = country_summary.sort_values('Total_sales',ascending=False).head()
top_five_countries

,Country,Number_of_transactions,Total_sales,Average_sales
36,United Kingdom,479985,9001744.09,18.75
24,Netherlands,2359,285446.34,121.00
10,EIRE,7879,283140.52,35.94
14,Germany,9025,228678.40,25.34
13,France,8392,209625.37,24.98


##Country that has the highest average transaction value

In [51]:
country_with_highest_avg = country_summary.sort_values('Average_sales',ascending=False).head(1)
country_with_highest_avg

,Country,Number_of_transactions,Total_sales,Average_sales
24,Netherlands,2359,285446.34,121.0


##Number of Transactions of the top-selling country

In [37]:
Top_country = country_summary.sort_values('Total_sales',ascending=False).head(1)
Top_country

,Country,Number_of_transactions,Total_sales,Average_sales
36,United Kingdom,479985,9001744.09,18.75


In [38]:
Top_country['Number_of_transactions'].iloc[0]

np.int64(479985)

1. The United Kingdom generated the highest total sales, mainly due to its very large number of transactions (479,985), despite having a lower average transaction value of 18.75.
2. The Netherlands had the highest average transaction value (121.00), showing that a high total sales value does not necessarily mean a high average transaction value.

Overall, the results demonstrate that a country's total sales are strongly influenced by transaction volume, whereas average sales reflect the typical value of an individual transaction. Therefore, a country can have a lower total sales value but a much higher average transaction value, as observed for the Netherlands.

#Part - 6 Merge and Region analysis

In [39]:
region_lookup = {
"United Kingdom": "Europe",
"France": "Europe",
"Germany": "Europe",
"Spain": "Europe",
"Netherlands": "Europe",
"Belgium": "Europe",
"Switzerland": "Europe",
"Portugal": "Europe",
"Australia": "Oceania",
"Japan": "Asia",
"USA": "Americas",
}

In [40]:
region_df = pd.DataFrame(region_lookup.items(),columns=['Country','Region'])
region_df

,Country,Region
0,United Kingdom,Europe
1,France,Europe
2,Germany,Europe
3,Spain,Europe
4,Netherlands,Europe
5,Belgium,Europe
6,Switzerland,Europe
7,Portugal,Europe
8,Australia,Oceania
9,Japan,Asia


##Merge it onto country summary table using a left merge on Country.

In [41]:
country_region = country_summary.merge(
    region_df,
    on='Country',
    how='left'
)

In [42]:
country_region

,Country,Number_of_transactions,Total_sales,Average_sales,Region
0,Australia,1181,138453.81,117.23,Oceania
1,Austria,398,10198.68,25.62,NaN
2,Bahrain,18,754.14,41.90,NaN
3,Belgium,2031,41196.34,20.28,Europe
4,Brazil,32,1143.60,35.74,NaN
5,Canada,151,3666.38,24.28,NaN
6,Channel Islands,747,20440.54,27.36,NaN
7,Cyprus,603,13502.85,22.39,NaN
8,Czech Republic,25,826.74,33.07,NaN
9,Denmark,380,18955.34,49.88,NaN


##Filling the gaps in Region with "Other"

In [43]:
country_region["Region"]=country_region["Region"].fillna("Other")
country_region

,Country,Number_of_transactions,Total_sales,Average_sales,Region
0,Australia,1181,138453.81,117.23,Oceania
1,Austria,398,10198.68,25.62,Other
2,Bahrain,18,754.14,41.90,Other
3,Belgium,2031,41196.34,20.28,Europe
4,Brazil,32,1143.60,35.74,Other
5,Canada,151,3666.38,24.28,Other
6,Channel Islands,747,20440.54,27.36,Other
7,Cyprus,603,13502.85,22.39,Other
8,Czech Republic,25,826.74,33.07,Other
9,Denmark,380,18955.34,49.88,Other


##Group by Region and find the total sales per region.

In [44]:
region_sales = country_region.groupby('Region')['Total_sales'].sum()
region_sales

,Total_sales
Region,
Americas,3580.39
Asia,37416.37
Europe,9918999.75
Oceania,138453.81
Other,543660.48


##Which region generated the highest total sales?

In [52]:
region_sales.sort_values(ascending=False)


,Total_sales
Region,
Europe,9918999.75
Other,543660.48
Oceania,138453.81
Asia,37416.37
Americas,3580.39


Europe generated the highest total sales (9918999.75), mainly because several high-sales countries, including the United Kingdom, are part of this region

#Part - 7  Answer the questions

1. Which country generated the highest total sales?

Answer - **United Kingdom**	generated the highest total sales with **9001744.09**.

2. Which country had the highest average transaction value?

Answer - **Netherlands** had the highest average transaction value, at **121.00** per transaction. This is much higher than the UK's average of 18.75	, even though UK had the highest total sales.

3. Which region performed best?

Answer - **Europe** had performed the best and generated highest Total sales(9918999.75) among all the regions.

4. What is one interesting pattern you noticed in the data?

Answer - An interesting pattern is that the country with the highest total sales did not have the highest average transaction value. The UK had a very high transaction volume, while the Netherlands had fewer transactions but a much higher average transaction value of 121.00.

5. What is one data-quality problem you ran into, and how did you handle it?

Answer -
The dataset contained duplicate transactions, which could have affected the sales analysis. I removed the duplicates before performing the analysis to avoid counting the same transaction more than once. And also the region lookup table did not contain every country, so some countries had missing (NaN) region values after the merge. I handled this by replacing the missing values with "Other".

#Part - 8 Final Insights

* **One important finding:** The United Kingdom generated the highest total sales, mainly because it had a very large number of transactions.(479985)
* **One comparison:** In comparison, the Netherlands had a much higher average transaction value (121.00) than the UK (18.75), despite having far fewer transactions.
* **One observation about the data itself:** The sales data is highly skewed, with a few very large values influencing the overall distribution.
* **One limitation:** Duplicate transactions were removed before analysis, so the final results may differ from the original dataset but provide a more reliable view of the sales.
